insonmnia/miner/network/l2tp_store.go

```diff
@@ -0,0 +1,349 @@
+	log "github.com/noxiouz/zapctx/ctxlog"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+type l2tpState struct {
+	Aliases  map[string]string
+	Networks map[string]*l2tpNetwork
+	mu       sync.Mutex
+	ctx      context.Context
+	storage  store.Store
+}
+
+func newL2TPNetworkStore(ctx context.Context, path string) (*l2tpState, error) {
+	boltdb.Register()
+	var (
+		backend   = store.Backend(store.BOLTDB)
+		endpoints = []string{path}
+		config    = store.Config{Bucket: "sonm_l2tp_driver_state"}
+	)
+	storage, err := libkv.NewStore(backend, endpoints, &config)
+	if err != nil {
+		return nil, err
+	}
+
+	state := &l2tpState{
+		Aliases:  make(map[string]string),
+		Networks: make(map[string]*l2tpNetwork),
+		ctx:      ctx,
+		storage:  storage,
+	}
+
+	if err := state.load(); err != nil {
+		return nil, err
+	}
+
+	return state, nil
+}
+
+func (s *l2tpState) Lock() {
+	s.mu.Lock()
+}
+
+func (s *l2tpState) Unlock() {
+	s.mu.Unlock()
+}
+
+func (s *l2tpState) Sync() (err error) {
+	defer func() {
+		if err != nil {
+			log.G(s.ctx).Error("failed to sync l2tp state", zap.Error(err))
+		}
+	}()
+
+	marshaled, err := json.Marshal(s)
+	if err != nil {
+		return err
+	}
+
+	err = s.storage.Put("state", marshaled, &store.WriteOptions{})
+	return
+}
+
+func (s *l2tpState) AddNetwork(netID string, netInfo *l2tpNetwork) error {
+	if _, ok := s.Aliases[netID]; ok {
+		return errors.Errorf("network already exists: %s", netID)
+	}
+
+	s.Aliases[netID] = netID
+	s.Networks[netID] = netInfo
+
+	return nil
+}
+
+func (s *l2tpState) AddNetworkAlias(netID, alias string) error {
+	if _, ok := s.Aliases[netID]; !ok {
+		return errors.Errorf("network not found: %s", netID)
+	}
+
+	s.Aliases[alias] = netID
+
+	return nil
+}
+
+func (s *l2tpState) RemoveNetwork(netID string) error {
+	translatedID, ok := s.Aliases[netID]
+	if !ok {
+		return errors.Errorf("network not found: %s", netID)
+	}
+
+	delete(s.Networks, translatedID)
+
+	for alias, target := range s.Aliases {
+		if target == translatedID {
+			delete(s.Aliases, alias)
+		}
+	}
+
+	return nil
+}
+
+func (s *l2tpState) GetNetwork(netID string) (*l2tpNetwork, error) {
+	translatedID, ok := s.Aliases[netID]
+	if !ok {
+		return nil, errors.Errorf("network not found: %s", netID)
+	}
+
+	if netInfo, ok := s.Networks[translatedID]; ok {
+		return netInfo, nil
+	}
+
+	return nil, errors.Errorf("network not found: %s", netID)
+}
+
+func (s *l2tpState) GetNetworks() []*l2tpNetwork {
+	var out []*l2tpNetwork
+	for _, netInfo := range s.Networks {
+		out = append(out, netInfo)
+	}
+
+	return out
+}
+
+type l2tpNetwork struct {
+	ID          string
+	PoolID      string
+	Count       int
+	NetworkOpts *L2TPNetworkConfig
+	Store       *l2tpEndpointStore
+}
+
+func newL2tpNetwork(opts *L2TPNetworkConfig) *l2tpNetwork {
+	return &l2tpNetwork{
+		NetworkOpts: opts,
+		Store:       NewL2TPEndpointStore(),
+	}
+}
+
+func (n *l2tpNetwork) Setup() error {
+	n.PoolID = n.NetworkOpts.GetHash()
```

Comment:
why does pool id is bound to options? What if I want to run 2 containers on the same host, connected to the same l2tp server?